MetaMorpheus/EngineLayer/DIA/PrecursorFragmentGrouping/XicGroupingEngine.cs

```diff
@@ -40,21 +54,35 @@ public override IEnumerable<PrecursorFragmentsGroup> PrecursorFragmentGrouping(L
                     {
                         var precursor = precursors[i];
                         var fragmentsInRange = GetXicsInRange(apexSortedFragmentXics, precursor.ApexRT, ApexRTTolerance);
-                        var pfGroup = GroupFragmentsForOnePrecursor(precursor, fragmentsInRange, OverlapThreshold, CorrelationThreshold, MinFragmentCountForPfGroup);
+                        var pfGroup = GroupFragmentsForOnePrecursor(precursor, fragmentsInRange);
                         if (pfGroup != null)
                         {
                             lock (pfGroups)
                                 pfGroups.Add(pfGroup);
                         }
                     }
                 });
-            return pfGroups;
+
+            var pfGroupsList = pfGroups.ToList();
+
+            //Filter precursor-fragment pairs by rank if rank thresholds are set
+            FilterPfPairsByRank(pfGroupsList, PrecursorRankThreshold, FragmentRankThreshold);
+
+            //Remove groups with insufficient fragment pairs after filtering
+            pfGroupsList.RemoveAll(g => g.PFpairs.Count < MinFragmentCountForPfGroup);
```

Comment:
To save time, you could also put an 

```
if (fragmentsInRange.Count < MinFragmentCountForPfGroup)
         continue;
```

on line 58. There are several locations throughout this PR in which loop short-circuiting can save you some computations. I will not hold up approval to make these optimizations.